Hannah-Sten/TeXiFy-IDEA
# Full compatibility for Sumatra custom path

Fix #2737

#### Summary of additions and changes
Added full support when inputting a custom Sumatra path, so that it appears in PDF Viewer ComboBox, and making it work with Forward Search and the Configure Inverse Search.
1. Changed isSumatraAvailable in SumatraConversation to a singleton object that checks if Sumatra is available.
2. Added an InputVerifier on the TextField of Custom Sumatra Path that checks the validity of the path, and updates the value of isSumatraAvailable and updates the items in the PDF Viewer ComboBox.
3. Changed the availability in the InternalPdfViewer to a function so that it can be changed during execution.
4. Removed the lines checking if the Sumatra Custom Path was null in addOpenViewerListener so that the user can select which PDF Viewer to open, even if a Sumatra Custom Path is present.
5. Added support for inverse search action for custom SumatraPath
6. Added support for loading custom SumatraPath at startup
7. Redone update of the PdfViewerComboBox to eliminate errors
8. Changed SumatraConversation from a class to a singleton object to better match InternalPdfViewers and to don't allow multiple instances of SumatraConversation to be opened by multiple parts of the program (Sumatra link conversation is constant over the scope of the program).
9. Added a run function that doesn't wait for a return of the program executed (useful for setting up Inverse search)

In brief : when inputting a correct custom Sumatra path (valid path and that contains SumatraPDF.exe), after clicking away from the TextBox, Sumatra automatically appears in the ComboBox of PDF Viewer. And then, making it work with Forward Search and the Configure Inverse Search.


#### How to test this pull request

Open TeXiFy with only global PATH set to Sumatra (afterwards, do the same with registry keys)
Under LaTeX configuration, under "PDF viewer", "Sumatra" should appear.
Under Tools/LaTeX, option to "Configure Inverse Search" should work as normal.
Forward search should work as normal.

Set no PATH and no registry keys to Sumatra.
Set custom path to Sumatra in Run Configuration to real Sumatra install and click away from the text box.
Under "PDF viewer", "Sumatra" should appear.
Set custom path to Sumatra in Run Configuration to a real folder that doesn't contain Sumatra  and click away from the text box.
Under "PDF viewer", "Sumatra" shouldn't appear and an error popup should appear.
Set custom path to Sumatra in Run Configuration to a folder that doesn't exist and click away from the text box.
Under "PDF viewer", "Sumatra" shouldn't appear and an error popup should appear.

Select a "PDF viewer" of your choice under LaTeX configuration, under "PDF viewer".
Run LaTeX project, the PDF viewer selected should open (even if Sumatra is in the Path).
